src/MarkdownTextInput.web.tsx

```diff
@@ -336,14 +255,60 @@ const MarkdownTextInput = React.forwardRef<TextInput, MarkdownTextInputProps>(
       }
     }, [multiline, onContentSizeChange]);
 
+    const parseInnerHTMLToText = useCallback((target: HTMLElement): string => {
+      let text = '';
+      const childNodes = target.childNodes ?? [];
+      childNodes.forEach((node, index) => {
+        const nodeCopy = node.cloneNode(true) as HTMLElement;
+        if (nodeCopy.innerHTML) {
+          // Replace single <br> created by contentEditable with '\n', to enable proper newline deletion on backspace, when next lines also have <br> tags
+          if (nodeCopy.innerHTML === '<br>') {
+            nodeCopy.innerHTML = '\n';
+          }
+          // Replace only br tags with data-id attribute, because we know that were created by the web parser. We need to ignore tags created by contentEditable div
+          nodeCopy.innerHTML = nodeCopy.innerHTML.replaceAll(/<br data-id=.*?>/g, '\n');
+        }
+        let nodeText = nodeCopy.textContent ?? '';
+
+        // Remove unnecessary new lines from the end of the text
+        if (nodeText.length > 2 && nodeText[-3] !== '\n' && nodeText.slice(-2) === '\n\n') {
+          nodeText = nodeText.slice(0, -1);
+        }
+
+        // Last line specific handling
+        if (index === childNodes.length - 1) {
+          if (nodeText === '\n\n') {
+            // New line creation
+            nodeText = '\n';
+          } else if (nodeText === '\n') {
+            // New line deletion on backspace
+            nodeText = '';
+          }
+        }
+
+        text += nodeText;
+        // Split paragraphs with new lines
+        if (/[^\n]/.test(nodeText) && index < childNodes.length - 1) {
+          text += '\n';
+        }
+      });
+      return text;
+    }, []);
+
     const handleOnChangeText = useCallback(
       (e: SyntheticEvent<HTMLDivElement>) => {
         if (!divRef.current || !(e.target instanceof HTMLElement)) {
           return;
         }
-        const changedText = e.target.innerText;
+
+        const parsedText = parseInnerHTMLToText(e.target);
+        divRef.current.value = parsedText;
+
+        const tree = buildTree(divRef.current, parsedText);
+        divRef.current.tree = tree;
+
```

Comment:
We don't pass the object generated here, because first we generate a tree and than we append generated dom to input's innerHTML. Because of this order we are ending up with outdated markdown element references in tree and cursor position algorithms break. We need to rebuild the tree to get the latest refs to elements in dom structure